src/ffs/normalizer/header.ts

```diff
@@ -0,0 +1,83 @@
+import ipfs, { CID } from '../../ipfs'
+import { BasicLinks, Links, FileSystemVersion, Header, Metadata } from '../types'
+import basic from './basic'
+import { isDecodingError, DecodingError, LinkDoesNotExistError, ContentTypeMismatchError } from './errors'
+import { notNull, isString, mapObjAsync } from '../../common'
+import link from '../link'
+
+export const getValue = async <T>(linksOrCID: Links | CID, name: string, checkFn: (obj: any) => obj is T, key?: string): Promise<T | DecodingError> => {
+  if(typeof linksOrCID === "string") {
+    const links = await basic.getLinks(linksOrCID, key)
+    return getValueFromLinks(links, name, checkFn, key)
+  }
+  return getValueFromLinks(linksOrCID, name, checkFn, key)
+}
+
+export const getValueFromLinks = async <T>(links: Links, name: string, checkFn: (obj: any) => obj is T, key?: string): Promise<T | DecodingError> => {
+  const linkCID = links[name]?.cid
+  if(!linkCID) {
+    return new LinkDoesNotExistError(name)
+  }
+  const value = await ipfs.encoded.catAndDecode(linkCID, key)
+  return checkFn(value) ? value : new ContentTypeMismatchError(linkCID)
+}
+
+export const getChildKey = async (cid: CID, key: string): Promise<string> => {
+  const childKey = await getValue(cid, "key", isString, key)
+  if(isDecodingError(childKey)){
+    throw new Error(`Could not retrieve child key: ${cid}. ${childKey.toString()}`)
+  }
+  return childKey
+}
+
+export const put = async(index: CID, header: Header, key?: string): Promise<CID> => {
+  const withVersion = {
+    ...header,
+    version: FileSystemVersion.v1_0_0
+  }
+  const linksArr = await Promise.all(
+    Object.entries(withVersion).map(async ([name, val]) => {
+      if(val !== undefined){
+        const cid = await ipfs.encoded.add(val, key)
+        return { name, cid, isFile: true }
+      }
+      return null
+    })
+  )
+  linksArr.push({ name: 'index', cid: index, isFile: false })
+  const links = link.arrToMap(linksArr.filter(notNull))
```

Comment:
I guess these are heterogeneous lists?

Reply:
Ah better yet, filter the list before mapping it, then the list is homogeneous & we don't have to use type guards:

```ts
    Object.entries(withVersion)
      .filter(([_name, val]) => val !== undefined)
      .map(async ([name, val]) => {
        const cid = await ipfs.encoded.add(val, key)
        return { name, cid, isFile: true }
      })
````